web-api-db-ctrl/src/postgres/extern_migrations.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::future::Future;
+
+use anyhow::bail;
+use async_trait::async_trait;
+use derive_more::{From, Into};
+use sqlx::{Postgres, Transaction, Type};
+
+use crate::Error;
+
+/// Allows running migration like "once per tenant" tasks.
+#[async_trait(?Send)]
+pub(crate) trait ExternalMigrator {
+    async fn run_migration_if_needed(
+        &mut self,
+        name: &str,
```

Comment:
❓ should we make this a `&'static str`? they name will be fixed anyways for each migration which we add.

Reply:
I don't think this will stay static.

E.g. we currently don't have a proper old schema => new schema ES migration (e.g. when adding new fields like `tags`)

And one thing I considered doing to fix that and additionally make some other code more robust is to not just check the embedding size for compatibility but check for each property in the expected if it exists and  if not create on the fly a `add-property-<name>` migration which adds that.

We can't hard code that for properties we added in the past, as AFIK our existing tenants might have a mixed state of weather they have an explicit `tags` mapping or not (it's happens to be either-way fine for tags and parents, but the "happens to" part is a bit risky long term)